llvm/lib/Target/M68k/M68kInstrAtomics.td

```diff
@@ -36,17 +57,128 @@ class MxCASOp<bits<2> size_encoding, MxType type>
   let mayStore = 1;
 }
 
-def CAS8  : MxCASOp<0x1, MxType8d>;
-def CAS16 : MxCASOp<0x2, MxType16d>;
-def CAS32 : MxCASOp<0x3, MxType32d>;
+def CASARI8  : MxCASARIOp<0x1, MxType8d>;
+def CASARI16 : MxCASARIOp<0x2, MxType16d>;
+def CASARI32 : MxCASARIOp<0x3, MxType32d>;
+
+class MxCASARIDOp<bits<2> size_encoding, MxType type>
+    : MxInst<(outs type.ROp:$out),
+             (ins type.ROp:$dc, type.ROp:$du, !cast<MxMemOp>("MxARID"#type.Size):$mem),
+             "cas."#type.Prefix#" $dc, $du, $mem"> {
+  let Inst = (ascend
+                (descend 0b00001, size_encoding, 0b011, MxEncAddrMode_p<"mem">.EA),
+                (descend 0b0000000, (operand "$du", 3), 0b000, (operand "$dc", 3))
+              );
+  let Constraints = "$out = $dc";
+  let mayLoad = 1;
+  let mayStore = 1;
+}
+
+def CASARID8  : MxCASARIDOp<0x1, MxType8d>;
+def CASARID16 : MxCASARIDOp<0x2, MxType16d>;
+def CASARID32 : MxCASARIDOp<0x3, MxType32d>;
+
+class MxCASARIIOp<bits<2> size_encoding, MxType type>
```

Comment:
Do we have MC tests for these new instructions (I assume you already have their codegen tests)? If not, could you add them?

Ditto for other CAS variants below.

Reply:
Added a new test at `llvm/test/MC/M68k/Atomics/cas.s` which shows all of them except `ARD`, which I couldn't get to parse for some reason. I was getting an illegal operand error when I tried to add `cas.w %d4, %d5, %a3` as a test case. Am I expecting something to work incorrectly here, or is it likely an issue with the AsmParser?

Reply:
On further investigation, looks like PC-relative and `ARD` addressing modes are not legal for `CAS`, will have to lower these to some different instruction sequences

Reply:
Removed the offending modes for now, as it looks like the code changes will be in different files. All of the addressing modes added by the patch are now tested in the new `MC` test